server/src/main/java/org/apache/kafka/server/DefaultAutoTopicCreationManager.java

```diff
@@ -0,0 +1,268 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server;
+
+import org.apache.kafka.clients.ClientResponse;
+import org.apache.kafka.common.errors.InvalidTopicException;
+import org.apache.kafka.common.internals.Topic;
+import org.apache.kafka.common.message.CreateTopicsRequestData;
+import org.apache.kafka.common.message.CreateTopicsRequestData.CreatableTopic;
+import org.apache.kafka.common.message.CreateTopicsRequestData.CreatableTopicConfig;
+import org.apache.kafka.common.message.CreateTopicsRequestData.CreatableTopicConfigCollection;
+import org.apache.kafka.common.message.MetadataResponseData.MetadataResponseTopic;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.AbstractRequest;
+import org.apache.kafka.common.requests.CreateTopicsRequest;
+import org.apache.kafka.common.requests.CreateTopicsResponse;
+import org.apache.kafka.common.requests.RequestContext;
+import org.apache.kafka.common.requests.RequestHeader;
+import org.apache.kafka.coordinator.group.GroupCoordinator;
+import org.apache.kafka.coordinator.group.GroupCoordinatorConfig;
+import org.apache.kafka.coordinator.share.ShareCoordinator;
+import org.apache.kafka.coordinator.share.ShareCoordinatorConfig;
+import org.apache.kafka.coordinator.transaction.TransactionLogConfig;
+import org.apache.kafka.server.common.ControllerRequestCompletionHandler;
+import org.apache.kafka.server.common.NodeToControllerChannelManager;
+import org.apache.kafka.server.config.AbstractKafkaConfig;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class DefaultAutoTopicCreationManager implements AutoTopicCreationManager {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAutoTopicCreationManager.class);
+
+    private final AbstractKafkaConfig config;
+    private final NodeToControllerChannelManager channelManager;
+    private final GroupCoordinator groupCoordinator;
+    private final ShareCoordinator shareCoordinator;
+    private final Supplier<Properties> transactionTopicConfigsSupplier;
+    private final Set<String> inflightTopics = ConcurrentHashMap.newKeySet();
+
+    public DefaultAutoTopicCreationManager(
+            AbstractKafkaConfig config,
+            NodeToControllerChannelManager channelManager,
+            GroupCoordinator groupCoordinator,
+            Supplier<Properties> transactionTopicConfigsSupplier,
+            ShareCoordinator shareCoordinator
+    ) {
+        this.config = config;
+        this.channelManager = channelManager;
+        this.groupCoordinator = groupCoordinator;
```

Comment:
it seems `groupCoordinator` could be replaced by a function, which returns `properties`